_posts/2023-04-14-kafka-transactions.md

```diff
@@ -0,0 +1,340 @@
+They are primarily used by the Kafka brokers themselves to manage topics, partitions, and other metadata, but may sometimes be used by administrators or monitoring tools as well.
+Control records are typically not visible to Kafka consumers, as they are meant to be processed only by the Kafka brokers.
+
+An end transaction marker (`COMMIT` or `ABORT`) is a special control record that indicates the end of a transaction (decision).
+It is used to ensure that the transaction is either completed successfully or aborted in case of a failure.
+The end transaction marker is added to the log after all the records related to the transaction have been written to the log.
+
+Non-transactional records are considered decided immediately, but transactional records are only decided when the corresponding commit or abort marker is written.
+
+A transaction goes through the following states:
+
+1. **Undecided** (Ongoing)
+2. **Decided and unreplicated** (PrepareCommit\|PrepareAbort)
+3. **Decided and replicated** (CompleteCommit\|CompleteAbort)
+
+This is handled through a process similar to the two-phase commit protocol.
+Every time some data is sent to a new partition inside a transaction, we have a new `Ongoing` control record.
+When the transaction completes, we have the `PrepareCommit` or `PrepareAbort` control record and the transaction coordinator is doomed to commit or abort.
+The final step consists of writing the end transaction markers to each involved partition, and the `CompleteCommit` or `CompleteAbort` control record.
+
+![txn log](/assets/images/posts/2023-04-14-kafka-txn-log.png)
+
+The last stable offset (LSO) is the offset such that all lower offsets have been decided and it is always present.
+The first unstable offset (FUO) is the earliest offset that is part of the ongoing transaction, if present.
+
+The LSO is equal to the FUO if it's lower than the HW, otherwise it's the HW.
+We have that `LSO <= HW <= LEO`, which can be the same offset when there is no open transaction and all partitions are in-sync.
+
+#### Error handling
+
+The `TransactionalIdAuthorizationException` happens when you have authorization enabled without the appropriate ACLs rules.
+
+In the following example we allow any `transactional.id` from any user.
+
+```sh
+$ $KAFKA_HOME/bin/kafka-acls.sh --bootstrap-server :9092 --command-config client.properties --add \
+  --allow-principal User:* --operation write --transactional-id *
+```
+
+If any of the send operations inside the transaction scope fails, the final `commitTransaction` will fail and throw the exception from the last failed send.
+When this happens, your application should call `abortTransaction` to reset the state and apply a retry strategy.
+Some errors cannot be recovered (`ProducerFencedException`, `FencedInstanceIdException`, `OutOfOrderSequenceException`) and the application should close the current producer instance and create a new one.
```

Comment:
How does the reader know which?

Reply:
I think you need to set a producer callback and log in case of error.